gson/src/main/java/com/google/gson/TypeAdapter.java

```diff
@@ -289,24 +289,29 @@ public final T fromJsonTree(JsonElement jsonTree) {
    * Note that we didn't need to check for nulls in our type adapter after we used nullSafe.
    */
   public final TypeAdapter<T> nullSafe() {
-    return new TypeAdapter<T>() {
-      @Override
-      public void write(JsonWriter out, T value) throws IOException {
-        if (value == null) {
-          out.nullValue();
-        } else {
-          TypeAdapter.this.write(out, value);
-        }
+    if (!NullSafeTypeAdapter.class.isInstance(this)) {
```

Comment:
Minor, but why not `!(this instanceof NullSafeTypeAdapter)`?

Reply:
@Marcono1234 

Yeah, I had `instanceof` in mind, but not sure how it all works as of today. The rule `org.apache.maven.enforcer.rules.version.RequireJavaVersion` fails with

```
Detected JDK version 21 (JAVA_HOME=/opt/.bin/jdk-21) is not in the allowed range [11,18)
```

If the type check is changed to `!(this instanceof NullSafeTypeAdapter)`, then:

* for JDK 11, it fails with `/src/main/java/com/google/gson/TypeAdapter.java:[292,27] illegal generic type for instanceof`;
* for JDK 17, it fails with `/src/main/java/com/google/gson/TypeAdapter.java:[292,16] reifiable types in instanceof are not supported in -source 7`.

The `Class.instanceOf` check works for all. However, according to the source code, it can work with `TypeAdapter<?>` at least in `GsonBuilder`.

 Please suggest.

Reply:
Ah sorry, it seems `!(this instanceof NullSafeTypeAdapter)` is not possible because `NullSafeTypeAdapter` is a non-static inner class (and the enclosing class `TypeAdapter<T>` is generic?).

The compiler error message is really not helpful, but in Eclipse IDE the error message included the suggested fix, qualify with the enclosing type: `TypeAdapter.NullSafeTypeAdapter`. Can you please try this instead?
```suggestion
    if (!(this instanceof TypeAdapter.NullSafeTypeAdapter)) {
```

Reply:
I've never encountered this error before preferring static inner classes to capture as few as possible, and now I see there are still "good-to-knows". But no doubt, `isAssignable` looked really weird and unnatural. Thank you!